src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -3020,6 +3020,119 @@ def cyclic_right_subideals(self, p, alpha=None):
             ans.append(J)
         return ans
 
+    
+    def is_integral(self):
+        r"""
+            Check if a quaternion fractional ideal is integral. An ideal in a quaternion algebra is said integral if it is contained in its left order. If the left order is already defined it just check the definition, otherwise it uses one of the alternative definition of Lemma 16.2.8 of [Voi2021]_.
+
+            Returns:
+                bool: True if the quaternion fractional ideal is integral, False otherwise.
+
+
+            EXAMPLES:
+
+            sage: R.<i,j,k> = QuaternionAlgebra(QQ, -1,-11)
+            sage: I = R.ideal([2 + 2*j + 140*k, 2*i + 4*j + 150*k, 8*j + 104*k, 152*k])
+            sage: I.is_integral()
+            True 
+            sage: O = I.left_order()
+            sage: I.is_integral()
+            True 
+            sage: I = R.ideal([1/2 + 2*j + 140*k, 2*i + 4*j + 150*k, 8*j + 104*k, 152*k])
+            sage: I.is_integral()
+            False
+
+        """
+        if self.__left_order is not None:
+            return all([b in self.left_order() for b in self.basis()])
+        elif self.__right_order is not None:
+            return all([b in self.right_order() for b in self.basis()])
+        else:
+            self_square = self**2
+            return all([b in self for b in self_square.basis()])
```

Comment:
Each of these tests should be equivalent to one of the form `A.free_module() <= B.free_module()`. (See also #37113.)